bam_markdup.c

```diff
@@ -852,6 +853,195 @@ static int mark_duplicates(md_param_t *param, khash_t(duplicates) *dup_hash, bam
 }
 
 
+static inline int optical_retag(md_param_t *param, khash_t(duplicates) *dup_hash, bam1_t *b, int paired, long *optical_single, long *optical_pair) {
+    int ret = 0;
+    uint8_t *data;
+
+    // remove any existing do tag
+    if ((data = bam_aux_get(b, "dt")) != NULL) {
+        bam_aux_del(b, data);
+    }
+
+    if (bam_aux_append(b, "dt", 'Z', 3, (const uint8_t *)"SQ")) {
+        fprintf(stderr, "[markdup] error: unable to append 'dt' tag.\n");
+        ret = -1;
+    }
+
+    if (paired) {
+        (*optical_pair)++;
+    } else {
+        (*optical_single)++;
+    }
+
+    if (param->supp) {
+        // Change the duplicate type
+
+        if (bam_aux_get(b, "SA") || (b->core.flag & BAM_FMUNMAP)
+            || bam_aux_get(b, "XA")) {
+            khiter_t d;
+
+            d = kh_get(duplicates, dup_hash, bam_get_qname(b));
+
+            if (d == kh_end(dup_hash)) {
+                // error, name should already be in dup hash
+                fprintf(stderr, "[markdup] error: duplicate name %s not found in hash.\n",
+                    bam_get_qname(b));
+                ret = -1;
+            } else {
+                kh_value(dup_hash, d).type = 'O';
+            }
+        }
+    }
+
+    return ret;
+}
+
+
+
+/*
+    Where there is more than one duplicate go down the list and check for optical duplicates and change
+    do tags (where used) to point to original (non-duplicate) read.
+*/
+static int duplicate_chain_check(md_param_t *param, khash_t(duplicates) *dup_hash, read_queue_t *ori,
+             long *warn, long *optical_single, long *optical_pair) {
+    int ret = 0;
+    read_queue_t *current = ori->duplicate;
+    char *ori_name = bam_get_qname(ori->b);
+    int have_original = !(ori->b->core.flag & BAM_FDUP);
+    int ori_paired = (ori->b->core.flag & BAM_FPAIRED) && !(ori->b->core.flag & BAM_FMUNMAP);
+
+    while (current) {
+        int current_paired = (current->b->core.flag & BAM_FPAIRED) && !(current->b->core.flag & BAM_FMUNMAP);
+
+        if (param->tag && have_original) {
+            uint8_t *data;
+
+            // at this stage all duplicates should have a do tag
+            if ((data = bam_aux_get(current->b, "do")) != NULL) {
+                // see if we need to change the tag
+                char *old_name = bam_aux2Z(data);
+
+                if (old_name) {
+                    if (strcmp(old_name, ori_name) != 0) {
+                        bam_aux_del(current->b, data);
```

Comment:
Note there are bam_aux_update functions now, but I'm not convinced they're worth using yet in terms of efficiency as they look rather like delete and append anyway (but even more complex).

Something to consider, but also definitely not a show stopper.

Reply:
Okay, I think this is something to remember for next time.